text/3446-store.md

```diff
@@ -0,0 +1,998 @@
+
+
+##  Pointer vs Reference
+
+A previous incarnation of the API provided borrow-checking. That is, resolving a handle would yield a reference and
+appropriately borrow the store.
+
+This is tempting, but I deem it a mistake.
+
+Specifically:
+
+1.  A mutably borrowed store cannot allocate, deallocate, nor further resolve any other reference. This makes
+    implementing any recursive data-structure -- such as a tree -- quite a bit more challenging than it ought to be.
+2.  A reference requires a type, for a fat reference this means metadata. Requiring the metadata to be provided when
+    calling `resolve` precludes the use of thin handles, which are quite welcome in graphs of objects with a great
+    number of copies of each handle.
+
+And the final nail in the coffin, for me, is that borrow-checking is best provided at the _handle_ level, rather than
+at the store level. The `UniqueHandle` possible future extension, which is non-intrusive and can be implemented in
+3rd-party code:
+
+-   Borrows the (unique) handle mutably or immutably, ensuring no misgotten access is provided.
+-   Borrows the store immutably, ensuring it is not moved nor dropped, which would potentially invalidate the pointers.
+
+This solution is more flexible, and more minimalist, generally a good sign with regard to API design.
+
+
+##  Argument-less dangling method
+
+A previous version of the companion repository used an argument-less `Store::dangling` method.
+
+The main advantage is that no instance of `Store` is then necessary to create an associated dangling handle. The
+somewhat hidden cost, however, is that `Store` is then no longer dyn-safe.
+
+An intermediate solution to restore dyn-safety would be a where clause `Self: Sized`, but while this would indeed make
+`Store` dyn-safe, it would still result in only providing partial functionality. This seems clearly undesirable.
+
+In the absence of strong usecase for creating dangling handles with no instance of `Store`, it seems preferable to have
+`Store::dangling` take `&self` so that `dyn Store` may be fully functional.
+
+
+##  Adapter vs Blanket Implementation
+
+A previous version of the companion repository used an `AllocatorStore` adapter struct, instead of a blanket
+implementation.
+
+There does not seem to be any benefit to doing so, and it prevents using collections defined in terms of a `Store`
+with an `Allocator`, which would require wrapping all store-based collections in allocator-based adapters in the
+standard library... and duplicate their documentation. Pure overhead.
+
+
+##  Marker granularity
+
+As a reminder, there are 3 marker traits:
+
+-   `MultipleStore`: allows concurrent allocations from a single store.
+-   `StableStore`: ensures existing pointers remain valid across calls to the API methods.
+-   `PinningStore`: ensures existing pointers remain valid even across moves.
```

Comment:
Because the reason `StableStore` is separate from `PinningStore` is subtle, I'd suggest including an example of such a `Store` and/or what `StableStore` gives without `PinningStore` / what extra `PinningStore` requires/gives.

The important difference is that while it's perfectly fine for dropping (the last sharing clone of) the storage to invalidate allocated memory in the store, leaking/forgetting (any sharing clone of) a `PinningStore` must ensure that resolved pointers are *never* invalidated (until such a time as (all sharing clones of) the store is unforgotten and dropped). This is easily the case for any storages allocating static memory, but non-`'static`/borrowing storages cannot be `PinningStore`, because after their captured lifetime has expired, the borrowed state can be dropped/deallocated even if the storage was leaked/forgotten and never dropped.

What I don't quite get is what the benefit of requiring `StableStore` might be. The best I can imagine is the ability to convert to and give `*mut T` to downstream code unaware of storages, but without the ability to go back to the handle type, it doesn't seem beneficial. The stated difference of pinning being validity across moves of the storage seems like it results from a misunderstanding of the extent of the pinning requirement. Interestingly, I do "feel(?)" at least some slight potential benefit to `StableStore` if it were to include stability across moves of the storage without the pinning requirement, though I can't quite concretize what that might be, despite not feeling like there's a proper use for `StableStore` without the stability across storage moves.

Reply:
I may definitely have misconceptions about pinning.

I had not thought of non `'static` storages which would have been borrowing their memory area, to be honest, and the `PinningStore` wording doesn't account for that. I'll leave this issue unresolved for now so it's not forgotten.

Reply:
> I had not thought of non `'static` storages

Interesting, given that nonstatic storage is a huge motivator for custom allocators. "zero sized static references" to static storage and inline storage may be the "pretty" cases (in that they don't result in duplicating allocator handles just to achieve drop cleanup), but temporary storage is where the real interesting memory management techniques come from.

Reply:
Indeed, that you mention it, it seems completely obvious :) I don't use stack allocators much, though, so didn't think about it.